File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.dart
class Solution {
  int characterReplacement(String s, int k) {
    int l = 0;
    int max_window = 0;
    List<int> counts = List.filled(26, 0);

    for (int r = 0; r < s.length; ++r) {
      // compute the frequency of the s[r] char what ever it was
      counts[s.codeUnitAt(r) - 65]++;
      // the repeating chars smaller than k (the number of allowed changes)
      // so that means this window is inValid
      int max_count = counts.reduce((a, b) => a > b ? a : b);

      if (((r - l + 1) - max_count) > k) {
        // we removed the left char
        counts[s.codeUnitAt(l) - 65]--;
        l += 1;
      }

      max_window = max(max_window, (r - l + 1));
    }
    return max_window;
  }
}

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
from collections import defaultdict

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        l = 0
        max_count = 0 
        count = defaultdict(int)
        n = len(s)
        result = 0
        for r in range(n):
            count[s[r]] += 1
            max_count = max(max_count, count[s[r]])
            
            # if the difference between window length and
            # the max count of the current window 
            # is greater than k (thats means we need to shift the window)
            if (r - l + 1) - max_count > k :
                count[s[l]] -= 1
                l += 1
            
            result = max(result, r- l + 1)
        return result